torch/_functorch/aot_autograd.py

```diff
@@ -1837,12 +1865,12 @@ def functionalized_f_helper(*args):
                     # We found an input that had a (data-only) mutation.
                     # Since keep_input_mutations is set, we need to faithfully apply a copy_()
                     # so the compiler will see the input mutation in the graph.
-                    assert inpt_new is not inpt_old
-                    if meta.input_info[i].mutations_hidden_from_autograd:
-                        with torch.no_grad(), torch.autograd._unsafe_preserve_version_counter(inpt_old):
+                    if not trace_joint or (trace_joint and not inpt_old.requires_grad and not meta.input_info[i].requires_grad):
```

Comment:
hmm what's this extra check for?

Reply:
This is to check when the input is updated so that it would be a tensor with grad. For example:
```
x = torch.tensor(1, requires_grad=False)
y = torch.tensor(1, requires_grad=True)
x.mul_(y)
```